File: src/lib/types/worksheet.ts
// Worksheet-focused data structures for document-based LilyPond generation

export type WorksheetType = 
  | 'chord-naming'
  | 'interval-recognition'
  | 'scale-building'
  | 'rhythm-exercise'
  | 'note-identification';

export type EditableElementType = 
  | 'chord'
  | 'note'
  | 'rest'
  | 'text'
  | 'time-signature'
  | 'key-signature';

export interface EditableElement {
  id: string;
  type: EditableElementType;
  position: {
    measure: number;
    beat: number;
    voice?: number;
  };
  content: string; // LilyPond notation or text
  isAnswer: boolean; // Whether this is a question or answer
  isInteractive: boolean; // Whether user can edit this element
}

export interface WorksheetSection {
  id: string;
  title: string;
  instructions?: string;
  elements: EditableElement[];
  layout: {
    measuresPerSystem: number;
    systemsPerPage: number;
    clef: 'treble' | 'bass' | 'both';
    timeSignature?: string;
    keySignature?: string;
  };
}

export interface WorksheetConfig {
  id: string;
  title: string;
  subtitle?: string;
  type: WorksheetType;
  sections: WorksheetSection[];
  globalSettings: {
    paperSize: 'letter' | 'a4';
    orientation: 'portrait' | 'landscape';
    showAnswers: boolean;
    fontSize: number;
  };
}

// Template definitions for different worksheet types
export interface WorksheetTemplate {
  id: string;
  name: string;
  description: string;
  type: WorksheetType;
  defaultSections: Omit<WorksheetSection, 'id' | 'elements'>[];
  createSections: (params: any) => WorksheetSection[];
}

// Chord naming worksheet specific types
export interface ChordNamingParams {
  chords: Array<{
    root: string; // e.g., "C", "F#", "Bb"
    quality: 'major' | 'minor' | 'diminished' | 'augmented' | 'dominant7' | 'major7' | 'minor7';
    position: {
      measure: number;
      beat: number;
    };
    showAnswer: boolean;
  }>;
  instructions?: string;
  layout: {
    chordsPerLine: number;
    showStaffLines: boolean;
  };
}

export interface InteractiveElement {
  id: string;
  element_type: string;
  bounds: ElementBounds;
  data: any;
}

export interface ElementBounds {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface LilyPondDocument {
  version: string;
  content: string;
  metadata: {
    title: string;
    composer?: string;
    tagline?: string;
  };
}

File: src-tauri/src/types/worksheet.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum WorksheetType {
    ChordNaming,
    IntervalRecognition,
    ScaleBuilding,
    RhythmExercise,
    NoteIdentification,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum EditableElementType {
    Chord,
    Note,
    Rest,
    Text,
    TimeSignature,
    KeySignature,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EditableElement {
    pub id: String,
    pub element_type: EditableElementType,
    pub position: ElementPosition,
    pub content: String,
    pub is_answer: bool,
    pub is_interactive: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ElementPosition {
    pub measure: u32,
    pub beat: u32,
    pub voice: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorksheetSection {
    pub id: String,
    pub title: String,
    pub instructions: Option<String>,
    pub elements: Vec<EditableElement>,
    pub layout: WorksheetSectionLayout,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorksheetSectionLayout {
    pub measures_per_system: u32,
    pub systems_per_page: u32,
    pub clef: Clef,
    pub time_signature: Option<String>,
    pub key_signature: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Clef {
    Treble,
    Bass,
    Both,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorksheetConfig {
    pub id: String,
    pub title: String,
    pub subtitle: Option<String>,
    #[serde(rename = "worksheetType")]
    pub worksheet_type: WorksheetType,
    pub sections: Vec<WorksheetSection>,
    pub global_settings: WorksheetGlobalSettings,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorksheetGlobalSettings {
    #[serde(rename = "paperSize")]
    pub paper_size: PaperSize,
    pub orientation: Orientation,
    #[serde(rename = "showAnswers")]
    pub show_answers: bool,
    #[serde(rename = "fontSize")]
    pub font_size: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PaperSize {
    Letter,
    A4,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Orientation {
    Portrait,
    Landscape,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChordNamingParams {
    pub chords: Vec<ChordDefinition>,
    pub instructions: Option<String>,
    pub layout: ChordLayout,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChordDefinition {
    pub root: String,
    pub quality: ChordQuality,
    pub position: ElementPosition,
    #[serde(rename = "showAnswer")]
    pub show_answer: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ChordQuality {
    Major,
    Minor,
    Diminished,
    Augmented,
    #[serde(rename = "dominant7")]
    Dominant7,
    #[serde(rename = "major7")]
    Major7,
    #[serde(rename = "minor7")]
    Minor7,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChordLayout {
    #[serde(rename = "chordsPerLine")]
    pub chords_per_line: u32,
    #[serde(rename = "showStaffLines")]
    pub show_staff_lines: bool,
}